<launch>
    <!-- Choose visualization -->
    <arg name="rviz" default="true" />
    <arg name="rtabmap_viz" default="false" />
    
    <param name="use_sim_time" type="bool" value="True"/>
   
    <!-- Run the ROS package stereo_image_proc for image rectification -->
    <group ns="/costar_husky_sensor_config_1">
        <node pkg="nodelet" type="nodelet" name="stereo_nodelet"  args="manager"/>
        <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
            <remap from="left/image_raw"    to="left/image_raw_throttle_relay"/>
            <remap from="left/camera_info"  to="left/camera_info_throttle"/>
            <remap from="right/image_raw"   to="right/image_raw_throttle_relay"/>
            <remap from="right/camera_info" to="right/camera_info_throttle"/>
            <param name="disparity_range" value="128"/>
        </node>
    </group>
         
    <!-- Stereo Odometry -->
    <node pkg="rtabmap_odom" type="stereo_odometry" name="stereo_odometry" output="screen">
        <remap from="left/camera_info"  to="/stereo_camera/left/camera_info_throttle"/>
        <remap from="right/camera_info" to="/stereo_camera/right/camera_info_throttle"/>
        <remap from="odom"              to="/stereo_odometry"/>

        <param name="frame_id" type="string" value="base_footprint"/>
        <param name="odom_frame_id" type="string" value="odom"/>
        <param name="Odom/Strategy" type="string" value="0"/>  
        <param name="Odom/EstimationType" type="string" value="1"/> 
        <param name="Odom/MinInliers" type="string" value="10"/>
        <param name="Odom/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
        <param name="Odom/MaxDepth" type="string" value="10"/>
        <param name="OdomBow/NNDR" type="string" value="0.8"/>
        <param name="Odom/MaxFeatures" type="string" value="1000"/>
        <param name="Odom/FillInfoData" type="string" value="$(arg rtabmap_viz)"/>
        <param name="GFTT/MinDistance" type="string" value="10"/>
        <param name="GFTT/QualityLevel" type="string" value="0.00001"/> 
    </node>
    
    <group ns="rtabmap">   
        <!-- Visual SLAM -->
        <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="--delete_db_on_start">
            <param name="frame_id" type="string" value="base_footprint"/>
            <param name="subscribe_stereo" type="bool" value="true"/>
            <param name="subscribe_depth" type="bool" value="false"/>
            <remap from="left/camera_info" to="/costar_husky_sensor_config_1/left/camera_info_throttle"/>
            <remap from="right/camera_info" to="/costar_husky_sensor_config_1/right/camera_info_throttle"/>
            <remap from="odom" to="/stereo_odometry"/>

            <param name="queue_size" type="int" value="30"/>
            <!-- RTAB-Map's parameters -->
            <param name="Rtabmap/TimeThr" type="string" value="700"/>
            <param name="Rtabmap/DetectionRate" type="string" value="1"/>
            
            <!-- Updated parameter names -->
            <param name="Kp/MaxFeatures" type="string" value="200"/>
            <param name="Kp/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
            <param name="Kp/DetectorStrategy" type="string" value="0"/>   <!-- use SURF -->
            <param name="Kp/NNStrategy" type="string" value="1"/>         <!-- kdTree -->

            <param name="SURF/HessianThreshold" type="string" value="1000"/>

            <!-- Updated based on new parameters -->
            <param name="Vis/EstimationType" type="string" value="1"/> <!-- 3D->2D (PnP) -->
            <param name="Vis/MinInliers" type="string" value="10"/>
            <param name="RGBD/LoopClosureReextractFeatures" type="string" value="true"/>
            <param name="Vis/MaxFeatures" type="string" value="500"/>
            <param name="Vis/MaxDepth" type="string" value="10"/> 
        </node>
      
        <!-- Visualisation RTAB-Map -->
        <node if="$(arg rtabmap_viz)" pkg="rtabmap_viz" type="rtabmap_viz" name="rtabmap_viz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
            <param name="subscribe_stereo" type="bool" value="true"/>
            <param name="subscribe_odom_info" type="bool" value="true"/>
            <param name="queue_size" type="int" value="10"/>
            <param name="frame_id" type="string" value="base_footprint"/>
            <remap from="left/image_rect" to="/costar_husky_sensor_config_1/left/image_rect_color"/>
            <remap from="right/image_rect" to="/costar_husky_sensor_config_1/right/image_rect"/>
            <remap from="left/camera_info" to="/costar_husky_sensor_config_1/left/camera_info_throttle"/>
            <remap from="right/camera_info" to="/costar_husky_sensor_config_1/right/camera_info_throttle"/>
            <remap from="odom_info" to="/odom_info"/>
            <remap from="odom" to="/stereo_odometry"/>
            <remap from="mapData" to="mapData"/>
        </node>
         
    </group>
  
   <!-- Visualisation RVIZ --> 
   <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap_demos)/launch/config/demo_stereo_outdoor.rviz"/>

</launch>
